functorch/_src/aot_autograd.py

```diff
@@ -131,29 +136,38 @@ def posthook(grad_input, grad_output):
         )
         node.register_hook(get_posthook(special_stack))
 
+# This class encapsulates all aliasing + mutation info we need about the forward graph
+# See a more detailed overview of the edge case handling at
+# https://docs.google.com/document/d/19UoIh_SVrMy_b2Sx5ZaeOJttm6P0Qmyss2rdBuyfoic/edit
+@dataclass(frozen=True)
+class ViewAndMutationMeta:
+    # length: # forward inputs
+    # For every input, tells us whether the input:
+    # (a) is not mutated
+    # (b) only metadata is mutated
+    # (c) data (and maybe metadta) is mutated
+    mutated_input_info: List[MutationType]
+    # length: (# inputs that have *only* metadata mutations)
+    # We stash the updated FakeTensor that we traced with in the forward in here,
+    # that way we can use it to replay the metadata mutation
+    metadata_mutation_input_info: List[Tensor]
+    # length: # outputs in the compiled forward. Equal to:
+    # length: (# inputs w data mutations) + (# outputs that don't alias inputs)
+    # For every output *and* mutated input returned from the forward,
+    # tells us whether or not the output should require gradients or not
+    requires_grad_out_info: List[bool]
+    # length: # outputs in the *outer* user's forward.
+    # For every output that is a view of an input, gives us information on how to recompute it later.
+    # If the i'th output aliases an input, this will be a tuple of
+    # (
+    #  idx of the input that this output aliases,
+    #   the FakeTensor representing that output, which we use to generate an as_strided() call with that metadata
+    # )
+    aliased_output_info: List[Optional[Union[int, torch.Tensor]]]
```

Comment:
s/Union/Tuple/

is mypy running on this file 🤔

Reply:
mypy is not running on this file :( it probably should...